finality-aleph/src/justification.rs

```diff
@@ -1,70 +1,220 @@
+
+    pub(crate) fn handle_requests(&mut self, num: NumberFor<B>) {
+        let current_time = Instant::now();
+        if current_time - self.last_finalization > Duration::from_secs(10)
+            && current_time - self.last_request > Duration::from_secs(10)
+        {
+            self.last_request = current_time;
+            debug!(target: "afa", "Trying to request block {:?}", num);
+            if let Ok(Some(header)) = self.client.header(BlockId::Number(num)) {
+                debug!(target: "afa", "We have block {:?} with hash {:?}. Requesting justification.", num, header.hash());
+                self.network
+                    .request_justification(&header.hash(), *header.number());
+            } else {
+                debug!(target: "afa", "Cancelling request, because we don't have block {:?}.", num);
+            }
         }
     }
 
     pub(crate) async fn run(mut self) {
-        while let Some(notification) = self.justification_rx.next().await {
-            self.finalization_proposals_tx
-                .unbounded_send(notification)
-                .expect("Notification should succeed");
+        loop {
+            let last_finalized_number = self.client.info().finalized_number;
+            let current_session = self.current_session_id(last_finalized_number);
+            let stop_h = (current_session + 1) * self.session_period - 1;
+            let keybox = self.session_keybox(current_session);
+            if keybox.is_none() {
+                error!(target: "afa", "Cannot fetch the keybox for session {:?}. Waiting 500ms and will try again ...", current_session);
+                Delay::new(Duration::from_millis(500)).await;
+                continue;
+            }
+            let keybox = keybox.expect("We loop until this is some.");
+            let mut ticker_fut = Delay::new(Duration::from_millis(1000));
+            tokio::select! {
+                maybe_notification = self.import_justification_rx.next() => {
+                    if let Some(notification) = maybe_notification {
+                        debug!(target: "afa", "Got justification (block import)");
+                        self.handle_justification_notification(notification, keybox, last_finalized_number, stop_h.into());
+                    }
+                }
+                maybe_notification= self.authority_justification_rx.next() => {
+                    if let Some(notification) = maybe_notification {
+                        debug!(target: "afa", "Got justification (local aggregator)");
+                        self.handle_justification_notification(notification, keybox,  last_finalized_number, stop_h.into());
+                    } else {
+                        error!(target: "afa", "Got a None from justification aggregator stream.");
+                    }
+                }
+                _ = &mut ticker_fut => {
+                    continue;
+                }
+            }
+            self.handle_requests(stop_h.into());
```

Comment:
Yeah, I was thinking about that and the short answer is: I'm pretty sure it is enough for this to work "reasonably" well.

By requesting `last_finalized_number + 1` (and having some delays between consecutive requests) we will be catching up very slowly so that might not be the best idea. The current vesion is minimal and should be enough to prevent disasters. If the session time is say ~10mins, then I would say it is reasonable.